### ISSUE TITLE
sam package returns error on windows while uploading to S3 because file is in use.

### ISSUE THREAD

**Issue by ctaylor-osv** [aws/aws-sam-cli#2270]
### Description

We have multiple lambdas in our SAM template and many of them are referencing the same local folder for the `CodeUri` setting. During the `sam package` command in the upload to S3 phase, we occasionally will get a `[WinError 32] The process cannot access the file because it is being used by another process` error. It does not happen every time, but typically 1 in 10 package statements fail. Usually running package again resolves the issue.

Previously, we had been using the command `aws cloudformation package` to package our SAM templates. However, we wanted to take advantage of the new SAM functionality of having a StateMachine's definition specified in a separate local file and found that the `aws cloudformation package` command was not appropriately packaging the separate definition. In order to get that to work, we switched our deployment pipeline to use `sam package`. Since that switch, we have been seeing these random deployment failures.

### Steps to reproduce

1. Create multiple lambdas in a SAM template
2. Set each of the lambdas 'CodeUri' values to the same local folder
3. Run `sam package` on the template

Reminder that I am not able to get it to regularly fail every time.

### Observed result

As this is ran as part of a build process and I am not able to regularly get it to fail, I haven't been able to capture the `--debug` output but here is the regular logs:
```
2020-10-02T11:20:17.0220992Z Uploading to 2aee04ade2e114c9c5a6ebcfab885b53  262144 / 303787.0  (86.29%)
2020-10-02T11:20:17.0895462Z Uploading to 2aee04ade2e114c9c5a6ebcfab885b53  303787 / 303787.0  (100.00%)
2020-10-02T11:20:17.0895885Z 
2020-10-02T11:20:17.2693631Z Uploading to fccf19a16d6bb3762777e5af7b1e769d  29325 / 29325.0  (100.00%)
2020-10-02T11:20:17.2694020Z 
2020-10-02T11:20:17.3653478Z Uploading to 6774fc7efe194a48b11cc9f41c57d770  262144 / 303600.0  (86.35%)
2020-10-02T11:20:17.6952610Z Uploading to 6774fc7efe194a48b11cc9f41c57d770  303600 / 303600.0  (100.00%)
2020-10-02T11:20:17.6953586Z Error: Unable to upload artifact ../dist/folder referenced by CodeUri parameter of MyFunction resource.
2020-10-02T11:20:17.6954429Z [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Windows\\SERVIC~2\\NETWOR~1\\AppData\\Local\\Temp\\data-6774fc7efe194a48b11cc9f41c57d770.zip'
```

### Expected result

I expect sam package to de-duplicate CodeUri parameter values and only upload each unique value to S3 once.

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: Windows Server 2016 Datacenter
2. `sam --version`: 1.2.0
3. NodeJS Lambda runtimes

`Add --debug flag to command you are running`

**Comment by forgivegod** [comment]
try upgrading to v1.4.0 
https://github.com/aws/aws-sam-cli/releases

**Comment by ctaylor-osv** [comment]
I upgraded to v1.4.0 and am still experiencing the problem. Permissions don't appear to be a problem because it works most of the time.

Regardless of whether `sam package` de-dups the `CodeUri` parameter or not, I previously had used `aws cloudformation package` command to package the same SAM template for years and never experienced this failure. Only since switching to the `sam package` command have I experienced this mode of failure. We switched to get the support for the `DefinitionUri` for a StateMachine to be able to point to a local file.

Any other ideas on how to resolve this? I'm going to go ahead and upgrade to v1.6.2 to so if it fixes this because it looks like in 1.6 there was a feature to de-dup builds (issue #1236). However, that looks like it is focused on `sam build`, not `sam package` so I'm somewhat doubtful. Having said that, it seems like if sam is de-duping the `build` command, asking for the `package` command to behave the same way seems reasonable.

**Comment by ctaylor-osv** [comment]
It failed again after upgrading to v1.6.2. This time it was a slightly different error, but still seems to be related to timing and multiple functions referencing the same `CodeUri` value:

```
2020-10-19T11:20:19.5349578Z Error: Unable to upload artifact ../dist/folder referenced by CodeUri parameter of MyFunction resource.
2020-10-19T11:20:19.5351114Z [WinError 2] The system cannot find the file specified: 'C:\\Windows\\SERVIC~2\\NETWOR~1\\AppData\\Local\\Temp\\data-3b1bd5ea03ed354ba7d17ea540d8c9e1.zip'
```

Again, a redeploy worked just fine.

**Comment by mndeveci** [comment]
Hi @ctaylor-osv ,

We tried to re-produce the issue locally with 3 code bases and each code base has 6 functions, in total 18 functions and 3 source code folders.

We ran the package command on OSX and Windows multiple times, but each time it succeeded without any error. 

We also tried to run multiple package commands at the same time (with multiple terminals open at the same project folder), then we saw the both errors (file not found and file is been used by another process). We believe that the root cause might be a race condition, if there are multiple `sam package` commands running for the same project.

Can you validate that each time only single `sam package` runs for the same project folder?

Thanks!

**Comment by ctaylor-osv** [comment]
Yes, there is only a single `sam package` execution for the same package folder. I've recently added the `--debug` flag to the package command so I'll see if I can provide any more details.

**Comment by mndeveci** [comment]
Thanks @ctaylor-osv ,

The other possibility which might cause this issue is, if you have same codebase in different folders, running `sam package` for both of them at the same time (because of same hashing) might be causing this issue as well.

**Comment by ctaylor-osv** [comment]
Thanks @mndeveci. I think this is it. We actually have deploys going out to multiple environments and they are running concurrently. I'm going to make one of the environments dependent on the other so that hopefully works around the issue.

We've been doing this for years, though, and have only been seeing the error since we started using `sam package` instead of `aws cloudformation package` so it still bothered me as to why there's a difference. I think I've identified what that is. In the [artifact_exporter.py for aws cli cloudformation command](https://github.com/aws/aws-cli/blob/55676cb3625dedf8f6c5036deca44a5c7815f229/awscli/customizations/cloudformation/artifact_exporter.py#L159-L177), the `zip_folder` method generates the filename using a `uuid`:
`filename = os.path.join(tempfile.gettempdir(), "data-" + uuid.uuid4().hex)`

In the [artifact_exporter.py for the sam cli](https://github.com/aws/aws-sam-cli/blob/7d5101a8edeb575b6925f9adecf28f47793c403c/samcli/lib/package/artifact_exporter.py#L177-L194), the `zip_folder` method generates the filename using a `md5hash`:
`filename = os.path.join(tempfile.gettempdir(), "data-" + md5hash)`

I understand the switch to the `md5hash` for uniqueness. However, this does introduce the issue I'm running into. Does it make sense to generate a `uuid` per execution of the `sam package` and create the zip in `filename = os.path.join(tempfile.gettempdir(), uuid, "data-" + md5hash)`?

**Comment by mndeveci** [comment]
I understand your concern @ctaylor-osv , but having filename with `data-{md5}` helps a lot if you don't make any change to the codebase. This is speeding up package command (during S3 upload) and prevent duplicated package uploads.

**Comment by ctaylor-osv** [comment]
@mndeveci I am not sure how what I am suggesting would change that. Currently, the `{tempdir}/data-{md5}` files are getting cleaned up every execution so it is not like it is saving time across executions. All I am suggesting is to create a sub-folder under temp per execution and still generating the file as `data-{md5}` so the resulting path is something like `{tempdir}/{execution-uuid}/data-{md5}`. Instead of using a uuid, it could even be easier to use the current process ID. This would lead to the path `{tempdir}/{pid}/data-{md5}`. This would still prevent duplicate package uploads while allowing for concurrent executions of the `sam package` command.

I am going to leave this issue open for now as I would like to get your feedback on the thoughts above. However, we are also looking at moving our `package` step earlier in the pipeline to help alleviate this issue. If you determine that it is not worth the effort to deal with this concurrency issue, feel free to go ahead and close the issue.

**Comment by mndeveci** [comment]
Released with [v1.82.0](https://github.com/aws/aws-sam-cli/releases/tag/v1.82.0)
